src/ipc/dbus/DBusObjectRegistry.cpp

```diff
@@ -78,30 +89,48 @@ void DBusObjectRegistry::unregisterObject(const QString &objectPath)
 {
     init();
     auto serviceName = DBusManager::instance().serviceName();
-    if (m_master) {
+    if (isMaster()) {
         m_master->unregisterObject(objectPath, serviceName);
     } else {
         return m_objectRegistryAsyncProxy->unregisterObject(objectPath, serviceName);
     }
 }
 
-const QMap<QString, QString> &DBusObjectRegistry::objects(bool blocking)
+QMap<QString, QString> DBusObjectRegistry::objects(bool blocking)
 {
     init();
-    if (m_master) {
-        return m_master->objects();
-    } else {
-        if (blocking) {
+    if (!isMaster() && blocking) {
+        if (m_objects.isEmpty()) {
             if (m_objectRegistryProxy == nullptr) {
-                m_objectRegistryProxy = new facelift::ipc::dbus::ObjectRegistryIPCDBusProxy(this);
+                m_objectRegistryProxy = new ObjectRegistryIPCDBusProxy(this);
                 m_objectRegistryProxy->ipc()->setServiceName(m_serviceName);
                 m_objectRegistryProxy->connectToServer();
             }
-            return m_objectRegistryProxy->objects();
-        } else {
-            return m_objectRegistryAsyncProxy->objects();
+            return m_objectRegistryProxy->getObjects();
         }
     }
+    return m_objects;
+}
+
+void DBusObjectRegistry::onObjectRegistered(const QString &objectPath, const QString &serviceName)
+{
+    if (!m_objects.contains(objectPath)) {
+        m_objects.insert(objectPath, serviceName);
+        emit objectsChanged();
+    }
+}
+
+void DBusObjectRegistry::onObjectRemoved(const QString &objectPath)
+{
+    if (m_objects.contains(objectPath)) {
+        m_objects.remove(objectPath);
+        emit objectsChanged();
+    }
```

Comment:
else ?

Reply:
Not needed. The purpose it to decide if signal should be emitted. I will reduce it: m_objects.contains(...) can be replaced with m_objects.remove(...).

Reply:
else {
    handleUnexpectedDataComingFromServer();
}

Reply:
?

Reply:
??

Already added qCCritical(LogIpc) << "Object does not exist. Object path:" << objectPath; as discussed.